[13.0] [FIX] QR-bill report layout

yvaucher opened the pull request:
### Description of the issue/feature this PR addresses:

Following #52517 (in v12)

QR-Bill was designed for a prior version of the specs:
https://www.paymentstandards.ch/dam/downloads/ig-qr-bill-en.pdf

This intend to reproduce more closely the example of the documentation and improve readability on the following aspects:

* Fix Reference :warning: this might change as it currently directly uses `l10n_ch_isr_number_spaced`
:hourglass_flowing_sand: on https://github.com/odoo/odoo/pull/52517 for a better fix
* Do not display info that are not filled
* Use font size as per the specs (max 10pt)
* Remove "Support" section (removed in v2.1)
* Fix position of amount
* Give enough space to avoid line break on the QRR
* Rename titles with simplified versions (missing translations)
* Regroup lines to improve readability
* Print the QR code without surrounding blank spaces to improve accuracy of placement and size
* Adds translations and set the report lang based on the customer. :switzerland: (:fr: :de: :it: )
* Adds thousand separators (blank spaces)
* Make it possible to generate an QRR without setting an ISR Subscription number, only the QR-IBAN is necessary.
* Adds much needed tests :muscle: 



### Current behavior before PR:


![2020-06-16-172636_1224x596_scrot](https://user-images.githubusercontent.com/4158438/84794478-954eb380-aff6-11ea-8afa-6447f49a982f.png)

With annotations:

![qr-bill-anotations](https://user-images.githubusercontent.com/4158438/84822912-cbebf480-b01d-11ea-8049-c38d3520c3eb.png)



### Desired behavior after PR is merged:

Here is the output with the content of this PR in 4 langs:

![2020-06-17-164520_1004x506_scrot](https://user-images.githubusercontent.com/4158438/84912843-2470cf80-b0ba-11ea-8259-b7f9c028edd1.png)
![2020-06-17-164532_1010x509_scrot](https://user-images.githubusercontent.com/4158438/84912906-35214580-b0ba-11ea-9cae-d4299cb3633a.png)
![2020-06-17-164547_1000x511_scrot](https://user-images.githubusercontent.com/4158438/84912886-2f2b6480-b0ba-11ea-937f-ca069b2b8d55.png)
![2020-06-17-164610_1032x553_scrot](https://user-images.githubusercontent.com/4158438/84912915-38b4cc80-b0ba-11ea-8c89-141d4cfd099c.png)



The aim is to look like the same as in Specs:

![2020-06-16-173156_770x430_scrot](https://user-images.githubusercontent.com/4158438/84795053-4ce3c580-aff7-11ea-982a-31a615284545.png)


TODO:

- [x] Translations
  - [x] Terms
  - [x] document based on customer lang
- [x] Receipt amount labels
- [x] "Acceptance point" position
- [x] QR is blurry?
- [x] QR placement
- [ ] Change to QR v25
- [x] fix thousand separators


EDIT: updated layout of QR code and Receipt

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

yvaucher commented on odoo/addons/base/models/ir_actions_report.py:
```diff
@@ -499,9 +499,13 @@ def barcode(self, barcode_type, value, width=600, height=100, humanreadable=0, q
         try:
             width, height, humanreadable, quiet = int(width), int(height), bool(int(humanreadable)), bool(int(quiet))
+            bar_border = 4
+            if barcode_type == 'QR' and quiet:
+                bar_border = 0
+
             barcode = createBarcodeDrawing(
                 barcode_type, value=value, format='png', width=width, height=height,
-                humanReadable=humanreadable, quiet=quiet
+                humanReadable=humanreadable, quiet=quiet, barBorder=bar_border
```
@KangOl @mart-e @smetl  your opinion on those few line is welcome. We need to be able to print a QR without the surrounding blank space to fix the `l10n_ch` QR code. Otherwise it's hard to make it 46mm x 46mm which is a requirement from the spec.

`quiet` is not supported by "QR" type. And is simlpy ignored.
But we can use `barBorder`.

You can check that at:
https://bitbucket.org/rptlab/reportlab/src/61ba11e7d1430641d97cbce044da6c8828ebac32/src/reportlab/graphics/barcode/qr.py#lines-86

Or do you prefer that we add a new keyword to `barcode`, my intent is to avoid changing signature on stable.

yvaucher replied:
BTW it would be a great improvement to also implement `barVersion`
https://bitbucket.org/rptlab/reportlab/src/61ba11e7d1430641d97cbce044da6c8828ebac32/src/reportlab/graphics/barcode/qr.py#lines-88

A using a higher version pushes back the limit of characters.

yvaucher commented:
@oco-odoo this PR is ready

It fixes numerous issues on the QR-bill.

I made it independent from the fixes on ISR and ISR-B

oco-odoo commented on odoo/addons/base/models/ir_actions_report.py:
```diff
@@ -499,9 +499,13 @@ def barcode(self, barcode_type, value, width=600, height=100, humanreadable=0, q
             width, height, humanreadable, quiet = int(width), int(height), bool(int(humanreadable)), bool(int(quiet))
+            bar_border = 4
+            if barcode_type == 'QR' and quiet:
+                bar_border = 0
```
Alright, but just add a small comment so that we understand that in case we need to change it in a distant future. Something like " 'quiet' argument is ignored by reportlab for QR-barcodes. The generation of a "quiet-zone" around the barcode is instead managed by 'barBorder' "

oco-odoo commented on addons/l10n_ch/models/res_bank.py:
```diff
@@ -184,6 +186,9 @@ def _is_qr_iban(self):
         different.
         """
+        if not self:
+            return False
```
Is there a legit case where we try to call it on an empty record ?

yvaucher replied:
This is to simplify the call to `_is_qr_iban` by replacing:

`self.invoice_partner_bank_id and self.invoice_partner_bank_id._is_qr_iban()`

by

`self.invoice_partner_bank_id._is_qr_iban()`

In master we might want to consider detecting a QR-IBAN by an other mean. Either with a computed field or as a separate type from IBAN.

oco-odoo commented on addons/l10n_ch/i18n_extra/it.po:
```diff
@@ -328,113 +328,48 @@ msgstr "910 Donazioni, dividendi, compensi, ..."
-#. module: l10n_ch
-#: model_terms:ir.ui.view,arch_db:l10n_ch.l10n_ch_swissqr_template
-msgid "<span class=\"swissqr_text title\">Additional information</span><br/>"
-msgstr "<span class=\"swissqr_text title\">Ulteriori informazioni</span><br/>"
```
@mart-e Is it allowed to remove translations like this in stable? Or should the new stuff just be put next to it?

mart-e replied:
@oco-odoo it's fine for `l10n_*` were translations only applies during the CoA installation. In stable it's problematic to change translatable content because the developer does not translate the new term in all languages on Transifex while merging the PR, here it is translated :)

yvaucher commented:
@oco-odoo I squashed the commits in one and added a comment in the code for your 2 remaining comments.

For the second one though, tell me if I have to replace any occurrence of `invoice.invoice_partner_bank_id._is_qr_iban()` by `invoice.invoice_partner_bank_id and invoice.invoice_partner_bank_id._is_qr_iban()` to make it a more realistic `api.multi`

Or do you prefer to make it a computed field? as not stored I guess it respects the no schema change in stable.

Any better option?

oco-odoo commented:
@yvaucher It's alright this way :) 
You squashed before committing your comments, so I have to ask you to squash again ^^ I'll merge as soon as it's done :)

yvaucher commented:
squashing now

yvaucher commented:
squashed

oco-odoo commented:
@robodoo r+
@yvaucher Thanks ! :)
